Parse

fernandobatels opened the pull request:
Implementação do parse do XML da NF-e.

**Tags**:

infNfe
- [x] versao
- [x] chave

ide
- [x] cUF
- [x] cNF
- [x] natOp
- [x] mod
- [x] serie
- [x] nNF
- [x] dhEmi
- [x] dhSaiEnt
- [x] tpNF
- [x] idDest
- [x] cMunFG
- [x] tpImp
- [x] tpEmis
- [x] cDV
- [x] tpAmb
- [x] finNFe
- [x] indFinal
- [x] indIntermed
- [x] indPres
- [x] procEmi
- [x] verProc

emit
- [x] CNPJ
- [x] xNome
- [x] xFant
- enderEmit
  - [x] xLgr
  - [x] nro
  - [x] xCpl
  - [x] xBairro
  - [x] cMun
  - [x] xMun
  - [x] UF
  - [x] CEP
  - [x] fone
- [x] IE
- [x] IEST

dest
- [x] CNPJ
- [x] xNome
- enderDest
  - [x] xLgr
  - [x] nro
  - [x] xCpl
  - [x] xBairro
  - [x] cMun
  - [x] xMun
  - [x] UF
  - [x] CEP
  - [x] fone
- [x] indIEDest
- [x] IE

det
- [x] nItem
- prod
   - [x] cProd
   - [x] cEAN
   - [x] xProd
   - [x] NCM
   - [x] CEST
   - [x] indEscala
   - [x] CNPJFab
   - [x] cBenef
   - [x] CFOP
   - [x] uCom
   - [x] qCom
   - [x] vUnCom
   - [x] vProd
   - [x] cEANTrib
   - [x] uTrib
   - [x] qTrib
   - [x] vUnTrib
   - [x] vFrete
   - [x] vSeg
   - [x] vDesc
   - [x] vOutro
   - [x] indTot
- imposto
   - [x] vTotTrib
   - ICMS
      - [x] ICMS60
      - [x] ICMSSN202
    - PIS
      - [x] PISOutr
      - [x] PISNT 
      - [x] PISAliq
    - COFINS
      - [x] COFINSNT
      - [x] COFINSOutr
      - [x] COFINSAliq

total
- ICMSTot
   - [x] vBC
   - [x] vICMS
   - [x] vOutro
   - [x] vNF
   - [x] vCOFINS
   - [x] vPIS
   - [x] vDesc
   - [x] vSeg
   - [x] vFrete
   - [x] vProd
   - [x] vTotTrib

transp
- [x] modFrete

infAdic
- [x] infCpl

fernandobatels committed:
Seção <ide> iniciada

- Material de apoio adicionado;
- Ajustes na seção <infNFE> e seus testes;
- Seção <ide> iniciada, ainda apenas com o cUF, cNF e natOP;

fernandobatels committed:
Número e série da NF-e

Campos número e série da NF-e mapeados

fernandobatels committed:
Atualização do material de apoio

fernandobatels committed:
Campo 'mod' mapeado

Campo 'mod', da tag <ide>, mapeado

fernandobatels committed:
Horário de emissão e saída mapeados

Campos 'emissao' e 'operacao' criados na struct 'Ide' para representarem
os campos dhEmi e dhSaiEnt.

fernandobatels committed:
Tipos e destinos de operação implementados

Tags 'tpNF' e 'idDest' mapeados para os campos tipo_operacao e
destino_operacao

fernandobatels committed:
Tag tpEmis mapeada

Campo tipo_emissao criado com base na tag tpEmis

fernandobatels committed:
Agrupamento de dados

Campos referentes a emissão e operação da nf, agrupados em duas novas
substructs dentro da Ide

fernandobatels committed:
Tag cMunFG mapeada

Campo codigo_municipio criado, sendo alimentado pela tag cMunFG

fernandobatels committed:
tpAmb e tpImp mapeados

Campos formato_danfe e ambiente adicionados a struct Ide, representando
o tipo de ambiente e de impressão da nfe

fernandobatels committed:
Ajustes enums

fernandobatels committed:
Consumidor e finalidade da emissão

Campo 'consumidor' adicionado ao struct Operacao, representando o valor
da tag 'indFinal' e campo 'finalidade' adicionado a struct Emissao,
representando a tag 'finNfe'

fernandobatels committed:
Campos presenca e intermediador criados

Campos presenca e intermediador criados na struct Operacao,
representando as tags indPres e indIntermed

fernandobatels committed:
Campos versao_processo e processo

Campos versao_processo e processo adicionados a struct Emissao,
representando as tags 'verProc' e 'procEmi'

fernandobatels committed:
Estrutura dos testes ajustada

fernandobatels committed:
Dígito verificador

Campo 'digito_verificador' criado, sendo alimentado pela tag 'cDV'

fernandobatels committed:
Parse do emitente

Struct para os dados do emitente criada. Parse do cnpj, razão social e
nome fantasia implementado.

fernandobatels committed:
IE e IEST

Campos ie e iest adicionados a struct Emitente, representando as tags IE
e IEST

fernandobatels committed:
Endereço do emitente

Tag 'enderEmit' mapeada para o campo 'endereco'

fernandobatels committed:
NFC-e

XML de teste para NFC-e adicionado. Testes ajustados para validarem o
novo formato.

Extraído do projeto nfephp-org/sped-nfe

fernandobatels committed:
Estrutura base para o destinatário criada

fernandobatels committed:
NfeBase x Nfe

Criado o conceito de NfeBase. Trata-se do tipo que abrangerá os dois
atuais modelos de notas(NFe e NFCe), deixando a cargo do tipo Nfe
apenas as tratativas referentes ao modelo 55(NFe).

Atualmente a principal diferença entre os modelos é a possível ausência
de um valor no campo destinatário no modelo 65(NFCe).

O usuário pode ainda fazer o parse da NFe diretamente de uma string ou
arquivo, mas se desejar acessar o tipo base, precisará usar o método
'parse' da crate.

Essa abordagem sobre os dois modelos ainda é experimental, sendo assim,
o modelo do NFCe foi deixado a cargo do usuário. Com o tempo, surgirá
alguma implementação que poderá fazer parte da crate.

fernandobatels committed:
Reestruturação dos módulos

Módulos da crate reorganizados, sendo divididos em:
- base/: módulos/tipos referentes a representação sem modelo definido;
- padrao/: módulos/tipos de uso comum a toda a crate;
- /: armazenamento dos módulos/tipos dos modelos já implementados;

fernandobatels committed:
padrao => comum

Módulo 'padrao' renomeado para 'comum', representando melhor o tipo do
seu conteúdo.

fernandobatels committed:
Melhorias no tipo Endereco

fernandobatels committed:
Parse do destinatário

Tipo Destinatário criado. Dado que alguns campos pode estar ausentes
em modelos diferentes do 55, o tipo foi implementado no módulo 'base' e
interfaceado no 'nfe'.

Módulos e tipos ajustados:
- base/nfe.rs passou a ser a raiz do módulo 'base';
- nfe.rs passou para o módulo 'nfe/mod.rs';

fernandobatels committed:
Destinatário: Indicador e IE

Campos indicador_ie e ie adicionados ao mapeamento do destinatário da NF

fernandobatels committed:
Estrutura simplificada

Estrutura da crate reorganizada. Agora apenas existem os módulos 'base'
e 'modelos'. Tudo o que for possível de se aproveitar nos modelos,
ficará neste primeiro. Já o segundo, contém os modelos de nfe
suportados.

fernandobatels committed:
TryFrom e FromStr para o base::Nfe

Traits TryFrom e FromStr implentadas diretamente na base da nfe. O
modelo 55 continuar implementando essas traits. Testes ajustados.

fernandobatels committed:
Produtos/Itens da nfe

Criadas as structs Item e Produto, sendo esta segunda já com os campos
código, descrição, gtin, cest e ncm

fernandobatels committed:
Suporte ao layout 4

Crate ajustada para apenas suportar, ao menos por enquanto, o
layout 4.

fernandobatels committed:
Benefício fiscal, cnpj fabricante e escala relevante

Campos código de benefício fiscal, cnpj fabricante e escala relevante
adicionados a struct de produtos

fernandobatels committed:
Quantidade, valor unitário, unidade, cfop e extipi

Campos quantidade, valor unitário, unidade, cfop e codigo_excecao_ipi
adicionados a struct Produto

fernandobatels committed:
Struct produto reorganizada

Struct ProdutoTributacao criada, sendo agora a responsável por fornecer
os dados referentes a tributação do produto.

fernandobatels committed:
Gtin, valor unitário, unidade e quantidade tributáveis

Campos gtin, valor unitário, unidade e quantidade tributáveis
incluídos na string ProdutoTributacao

fernandobatels committed:
Valores do produto adicionados

Campos referentes aos valores do produto adicionados na struct Produto

fernandobatels committed:
cargo fmt

fernandobatels committed:
Impostos dos itens

Estrutura para representar os impostos de cada item criada. No momento
apenas os grupos ICMS60 e ICMSSN202 estão implementados.

fernandobatels committed:
Grupo PIS

Grupos de impostos do PIS adicionados a estrutura de impostos dos
itens. No momento apenas os grupos PISNT, PISOutr e PISAliq estão
implementados.

fernandobatels committed:
Módulo base reestruturado

Módulo base reestruturado, contando agora com os submódulos 'ide' e
'item'. Dentro do 'item' há também uma nova subdivisão, o módulo
'imposto'.

fernandobatels committed:
Grupo COFINS

Grupos de impostos do COFINS adicionados a estrutura de impostos dos
itens. No momento apenas os grupos COFINSNT, COFINSOutr e COFINSAliq
estão implementados.

fernandobatels committed:
Totalização da nota

Struct Totalizacao criada, sendo a responsável pelo parse e acesso dos
dados da tag total>ICMSTot. A nova struct é disponibilizada no campo
'totais' da base da NFe.

fernandobatels committed:
Informações adicionais da NF-e

Tag 'infCpl' mapeada para o campo 'informacao_complementar' da struct
Nfe/NfeBase

fernandobatels committed:
Cargo fmt

fernandobatels committed:
Transporte da NF

Tag 'modFrete' mapeada para o campo 'modalidade' da nova struct
'Transporte'

fernandobatels commented:
Depois de muita reestruturação, creio que consegui sintetizar a minha ideia para a organização dos dados. 

Alguns dados ficaram de fora, pois não quero prolongar essa PR mais ainda, mas que podem/serão tratados futuramente:
- Dados sobre o pagamento;
- Dados sobre a cobrança;
- Dados sobre o IPI, na seção de impostos dos produtos;
- Suporte a serviços/ISS

otavio commented on nfe/src/base/dest.rs:
```diff
@@ -0,0 +1,87 @@
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub enum IndicadorContribuicaoIe {
+    /// Contribuinte ICMS
+    Contribuinte = 0,
+    /// Contribuinte isento de Inscrição no cadastro de Contribuintes
+    Isento = 1,
+    /// Não Contribuinte, que pode ou não possuir Inscrição Estadual no Cadastro de Contribuintes do ICMS
+    NaoContribuinte = 9,
```
O valor Isento esta diferente no `enum` e no `FromStr` abaixo.

otavio commented on nfe/src/base/dest.rs:
```diff
@@ -0,0 +1,87 @@
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "9" => IndicadorContribuicaoIe::NaoContribuinte,
+            "2" => IndicadorContribuicaoIe::Isento,
+            _ => IndicadorContribuicaoIe::Contribuinte, // 1
```
Eu sugiro usar uma linha para cada item do enum e o `_` usar um `unreacheable!`.

otavio commented on nfe/src/base/emit.rs:
```diff
@@ -0,0 +1,65 @@
+            .ok_or("Tag <xNome> não encontrada na <emit>")?[0]
+            .to_string();
+
+        let nome_fantasia = {
+            if let Some(fa) = parsercher::search_text_from_tag_children(&emit, &Tag::new("xFant")) {
+                Some(fa[0].to_string())
+            } else {
+                None
+            }
+        };
```
Acho que esse `if let` pode ser convertido para um `map`.

fernandobatels replied:
Sim, parece que pode ser convertido mesmo. Como vou tentar empregar o serde, não me preocuparei com essa questão agora.

otavio reviewed (changes requested):
Eu revisei o inicio da PR e vi umas possiveis melhorias. Da uma analisada. Todas elas parecem serem aplicaveis em mais locais.

Outro ponto eh que parece nao estar de acordo com o estilo de codigo.

fernandobatels committed:
Enums revisados

FromStr dos enums revisados, bem como empregado o uso da macro
unreachable!

fernandobatels commented:
Vou tirar a PR de revisão para tentar implementar o uso do serde no lugar do parse manual.

fernandobatels committed:
Convertendo para o serde

Iniciada a substituição do parse manual pelo uso do serde. As structs
Totalizacao e Transporte já suportam o novo formato. Todas as structs
de nível principal implentaram o FromStr, facilitando o parse individual
delas usando o serde.

No momento o serde está sendo implementado ao lado do parse que já
existe, mas assim que todas as structs estiverem convertidas, o parse
manual será removido.

fernandobatels committed:
Serde: destinatário, emitente e endereço

Suporte ao parse via serde implementado nas structs destinatário,
emitente e endereço.

fernandobatels committed:
Serde: ide

Suporte do parse via serde implementado para a seção 'ide' do xml

fernandobatels committed:
Serde: produto

Suporte do parse via serde implementado para as structs Produto e
ProdutoTributacao

fernandobatels committed:
Serde: impostos e itens

Suporte do parse via serde implementado para as structs Item e Imposto

fernandobatels committed:
Serde implementado

Base da NFe convertida para o uso do serde

fernandobatels committed:
Crate parsercher removida

Antiga implementação do parse, usando o parsercher, removida do
projeto. Agora todo o parse está baseado no uso da crate serde

fernandobatels committed:
Merge branch 'main' into parse
